### PR TITLE
[READY]Sparklers, Firecrackers and Ian's New year's party.

### PR DESCRIPTION
**Pull request by Krysonism**
## About The Pull Request

This PR introduces sparklers, firecrackers and adds them to the HoPs office during new years eve. 

The box for this event has a 20% chance to contain a frag grenade instead of a snap pop, for that little extra holiday magic. 
While this is significantly less powerful than the guaranteed revolver the janitor gets during garbage day, i am willing to replace the meme box with the normal version if the maintainers take issue with this.

The sparklers provide a little bit of light and can be used to ignite stuff once lit.

The firecrackers are very low yield explosives that must be ignited to start timing. 

You can shorten the fuse with a wirecutter.

Ian also gets a festive hat and some dog champagne.

firecracker64
sparkler_64

## Why It's Good For The Game

Ian's birthday has been pushed back 6 months, so it is only fair Ian gets to have a little bit of fun now.

The Ian's new years eve party is a little barebones since I've only have had a few hours to do this project, but it can be expanded upon when  we have time and good ideas.

## Changelog
:cl:
add: Added sparklers and firecrackers to the HoP's office during new years eve.
add: Ian can now wear the festive hat, one is provided to him at the end of the year.
/:cl:

### CODE REVIEW

File: code/game/objects/items/storage/boxes.dm [hunk 1]
```diff
@@ -1201,3 +1201,44 @@
...
+
+/obj/item/storage/box/fireworks
+	name = "box of fireworks"
+	desc = "Contains an assortment of fireworks."
+	illustration = "flashbang"
+
+/obj/item/storage/box/fireworks/PopulateContents()
+	for(var/i in 1 to 3)
+		new/obj/item/sparkler(src)
+	for(var/i in 1 to 3)
```
Comment (Cyberboss) [inline code/game/objects/items/storage/boxes.dm]:
```suggestion
```

File: code/game/objects/items/storage/boxes.dm [hunk 2]
```diff
@@ -1201,3 +1201,44 @@
...
+		new/obj/item/grenade/firecracker(src)
+	new /obj/item/toy/snappop(src)
+
+/obj/item/storage/box/fireworks/dangerous
+
+/obj/item/storage/box/fireworks/dangerous/PopulateContents()
+	for(var/i in 1 to 3)
+		new/obj/item/sparkler(src)
+	for(var/i in 1 to 3)
```
Comment (Cyberboss) [inline code/game/objects/items/storage/boxes.dm]:
```suggestion
```

File: code/modules/holiday/new_year.dm [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	light()
+
+/obj/item/sparkler/attackby(obj/item/W, mob/user, params)
+	var/ignition_msg = W.ignition_effect(src, user)
+	if(ignition_msg)
+		light(ignition_msg)
+	else
+		return ..()
+
+/obj/item/sparkler/proc/light(message)
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]:
```suggestion
/obj/item/sparkler/proc/light(mob/user, message)
```

File: code/modules/holiday/new_year.dm [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	var/lit = FALSE
+	heat = 1000
+
+/obj/item/sparkler/fire_act(exposed_temperature, exposed_volume)
+	light()
+
+/obj/item/sparkler/attackby(obj/item/W, mob/user, params)
+	var/ignition_msg = W.ignition_effect(src, user)
+	if(ignition_msg)
+		light(ignition_msg)
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]:
```suggestion
		light(user, ignition_msg)
```

File: code/modules/holiday/new_year.dm [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	else
+		return ..()
+
+/obj/item/sparkler/proc/light(message)
+	if(lit)
+		return
+	if(message)
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]:
```suggestion
	if(user && message)
```

File: code/modules/holiday/new_year.dm [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+		usr.visible_message(message)
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]:
```suggestion
		user.visible_message(message)
```

File: code/modules/holiday/new_year.dm [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	set_light(l_range = 2, l_power = 2)
+	damtype = "fire"
+	START_PROCESSING(SSobj, src)
+	playsound(src, 'sound/effects/fuse.ogg', 20, TRUE)
+	update_icon()
+
+/obj/item/sparkler/process()
+	burntime--
+	if(burntime < 1)
+		STOP_PROCESSING(SSobj, src)
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]: Put this line in Destroy()

File: code/modules/holiday/new_year.dm [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+/obj/item/grenade/firecracker/fire_act(exposed_temperature, exposed_volume)
+	prime()
+
+obj/item/grenade/firecracker/wirecutter_act(mob/living/user, obj/item/I)
+	if(active)
+		return
+	if(det_time)
+		det_time -= 10
+		to_chat(user, "<span class='notice'>You shorten the fuse of [src] with [I].")
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]: Maybe give some feedback on det_time here

File: code/modules/holiday/new_year.dm [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+		playsound(src, 'sound/items/wirecutter.ogg', 20, TRUE)
+		icon_state = initial(icon_state) + "_" + "[det_time]"
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]:
```suggestion
		icon_state = initial(icon_state) + "_[det_time]"
```

File: code/modules/holiday/new_year.dm [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	else
+		to_chat(user, "<span class='danger'>You've already removed all of the fuse!")
+
+obj/item/grenade/firecracker/preprime(mob/user, delayoverride, msg = TRUE, volume = 80)
+	var/turf/T = get_turf(src)
+	log_grenade(user, T)
+	if(user)
+		add_fingerprint(user)
+		if(msg)
+			to_chat(user, "<span class='warning'>You prime [src]! [capitalize(DisplayTimeText(det_time))]!</span>")
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]: Bad spot to display the detonation time imo. Probably should be in examine.

File: code/modules/holiday/new_year.dm [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	playsound(src, 'sound/effects/fuse.ogg', volume, TRUE)
+	active = TRUE
+	icon_state = initial(icon_state) + "_active"
+	addtimer(CALLBACK(src, .proc/prime), isnull(delayoverride)? det_time : delayoverride)
+
+/obj/item/grenade/firecracker/prime()
+	update_mob()
+	explosion(src.loc,-1,-1,2)
+	qdel(src)
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]: qdel before explosion

File: code/modules/holiday/new_year.dm [hunk 12]
```diff
@@ -0,0 +1,111 @@
+//~*~*~*~*SPARKLER*~*~*~*~*~*~
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]: This file doesn't belong here imo. I'd suggest `code/game/objects/items/grenades/festive.dm`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/holiday/new_year.dm [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+			to_chat(user, "<span class='warning'>You prime [src]! [capitalize(DisplayTimeText(det_time))]!</span>")
+	playsound(src, 'sound/effects/fuse.ogg', volume, TRUE)
+	active = TRUE
+	icon_state = initial(icon_state) + "_active"
+	addtimer(CALLBACK(src, .proc/prime), isnull(delayoverride)? det_time : delayoverride)
+
+/obj/item/grenade/firecracker/prime()
+	update_mob()
+	explosion(src.loc,-1,-1,2)
+	qdel(src)
```
Comment (Cyberboss) [inline code/modules/holiday/new_year.dm]: qdel before explosion

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/holiday/new_year.dm (hunk 8 above)
Comment (Krysonism) [reply code/modules/holiday/new_year.dm]: The sprite changes, I also think you can examine it.

**Comment by knacker48** [comment]
Any plans to make this available in other forms? Maybe a crate from cargo?

File: code/modules/holiday/new_year.dm (hunk 10 above)
Comment (Krysonism) [reply code/modules/holiday/new_year.dm]: This is in line with the other grenades. I think it is in examine as well.

**Comment by Krysonism** [comment]
> 
> 
> Any plans to make this available in other forms? Maybe a crate from cargo?

Yes, but i don't want the PR to be held up due to even the slightest balance concern.

I even made a variant of the fireworks box that has a 20% chance to replace the snap pop with a frag grenade.

 Should i spawn this variant for the HoP during new year's eve instead for memes? I mean during one holiday the janitor gets a fucking revolver, so a 20% grenade chance doesn't sounds so bad.

**Comment by Krysonism** [comment]
Putting qdel before explosion as instructed, does not work.

**Comment by OnlineGirlfriend** [comment]
Glad somebody is thinking of Ian.

**Comment by Cyberboss** [comment]
>Putting qdel before explosion as instructed, does not work.

Get the explosion location first:
```dm
var/explosion_loc = get_turf(src)
qdel(src)
explosion(explosion_loc, ...)
```

**Comment by Krysonism** [comment]
> 
> 
> > Putting qdel before explosion as instructed, does not work.
> 
> Get the explosion location first:
> 
> ```dm
> var/explosion_loc = get_turf(src)
> qdel(src)
> explosion(explosion_loc, ...)
> ```

Okay, will do.

Travis is telling me to run map merger, i will have to figure out how to do so in an hour or so once i get some food and rest.

File: code/modules/holiday/new_year.dm (hunk 13 above)
Comment (Krysonism) [reply code/modules/holiday/new_year.dm]: Fix this, thanks.

File: code/modules/holiday/new_year.dm (hunk 12 above)
Comment (Krysonism) [reply code/modules/holiday/new_year.dm]: I have implemented your suggestion, they are now under grenades. Probably a good choice if the number of fireworks should expand in the future.

File: code/modules/holiday/new_year.dm (hunk 11 above)
Comment (Krysonism) [reply code/modules/holiday/new_year.dm]: Fixed

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/objects/items/grenades/festive.dm [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+/obj/item/sparkler
+	name = "sparkler"
+	desc = "A little stick coated with metal power and barium nitrate, burns with a pleasing sparkle."
+	icon = 'icons/obj/holiday_misc.dmi'
+	icon_state = "sparkler"
+	w_class = WEIGHT_CLASS_TINY
+	var/burntime = 60
+	var/lit = FALSE
+	heat = 1000
```
Comment (Cyberboss) [inline code/game/objects/items/grenades/festive.dm]: Move this above the `var/` declarations

**Review by Cyberboss — DISMISSED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**